lib/block-write-stream.ts

```diff
@@ -56,10 +62,50 @@ export class BlockWriteStream extends Writable {
 			return;
 		}
 
-		this.__write(buffer, this.bytesWritten).then(callback, callback);
+		if (this._bytes === 0 && buffer.length >= this.chunkSize) {
+			if (buffer.length % this.blockSize === 0) {
+				this.__write(buffer, this.bytesWritten).then(callback, callback);
+				return;
+			}
+		}
+
+		this._buffers.push(buffer);
+		this._bytes += buffer.length;
+
+		if (this._bytes >= this.chunkSize) {
+			let block = Buffer.concat(this._buffers);
+			const length = Math.floor(block.length / this.blockSize) * this.blockSize;
+
+			this._buffers.length = 0;
+			this._bytes = 0;
+
+			if (block.length !== length) {
+				this._buffers.push(block.slice(length));
+				this._bytes += block.length - length;
+				block = block.slice(0, length);
+			}
+
+			this.__write(block, this.bytesWritten).then(callback, callback);
+			return;
+		}
+
+		callback();
 	}
 
 	async __final(): Promise<void> {
+		if (this._bytes) {
+			const length = Math.ceil(this._bytes / this.blockSize) * this.blockSize;
+			const block = Buffer.alloc(length);
+			let offset = 0;
+
+			for (let index = 0; index < this._buffers.length; index += 1) {
+				this._buffers[index].copy(block, offset);
+				offset += this._buffers[index].length;
+			}
+
+			await this.__write(block, this.bytesWritten);
```

Comment:
What happens with leftover bytes?

Reply:
I'm not sure I follow – this copies all remaining buffered chunks into `block`, and then writes those out, and as this is in `_final()`, there won't be any more writes to the stream after it

Reply:
It doesn't. If `this._bytes` is 1025 bytes and `this.blockSize` is 1024 then `block` will have a length of `2048` and you'll write 1023 bytes of garbage.

Reply:
Not garbage, but zeros; and that's the point of this, as we need to pad the last write to the device's block size – or am I missing something?

Reply:
I assume blockSize is a multiple of the device's block size to make this fast. Something like 4MB, right? In that case not only it's bad that we write more than asked for but you also risk hitting an write error if you cross the device end boundary. Why don't we just write out the buffer and let the OS do the last chunk?

Reply:
resolved in flowdock. Will be addressed in the future https://github.com/resin-io-modules/etcher-sdk/issues/19

Reply:
blockSize would be the device's block size, while chunkSize is the preferred minimum chunk size to write in one go, so we only write the minimum required additional zero-filled data.

> Why don't we just write out the buffer and let the OS do the last chunk?

Because Windows & MacOS require writes to block devices to match the device's block size (either physical or logical), which is 512 bytes (also for 4K devices, as they have to support the legacy 512B as well)